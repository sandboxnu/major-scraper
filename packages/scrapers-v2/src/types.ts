/**
 * Catalog Representations:
 *
 * { catalog { <courseList> ... } }
 * { courseList <desc> <comment> <courseBody> } - table
 * { courseBody { <courseRow | textRow> ...} }
 *    textRow:
 *      { areaHeader <comment> <hour> }
 *      { commentRow <comment> <hour> }
 *      { subHeader  <comment> <hour> }
 *
 *    courseRow:
 *      { courseRow <codehol> <hour> }
 *      { orCourseRow <codehol> <hour> }
 *
 */

export type HTMLCatalog = {
  yearVersion: number;
  majorName: string;
  prgramRequiredHours: number;
  courseLists: HTMLCatalogCourseList[];
};

export type HTMLCatalogCourseList = {
  description: string;
  courseBody: HTMLCatalogCourseListBodyRow[];
};

export interface HTMLCatalogCourseListBodyRow {
  description: string;
  hour: number;
  courseTitle?: string;
  type: CourseListBodyRowType;
}

export enum CourseListBodyRowType {
  HEADER = "HEADER",
  SUBHEADER = "SUBHEADER",
  COMMENT = "COMMENT",
  OR_COURSE = "OR_COURSE",
  PLAIN_COURSE = "PLAIN_COURSE",
}
