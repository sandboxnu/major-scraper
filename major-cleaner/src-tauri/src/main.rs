// Prevents additional console window on Windows in release, DO NOT REMOVE!!
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use std::fs;

// TODO: investigate strong typing for Tauri commands (maybe something like https://github.com/oscartbeaumont/tauri-specta ?).

#[tauri::command]
fn read_major_file(year: &str, college: &str, name: &str, form: &str) -> String {
    // TODO: ensure each path segment doesn't contain special path chars (ie. /, .., etc.)
    let path = format!("../../degrees/Major/{year}/{college}/{name}/{form}");
    match fs::read_to_string(&path) {
        Ok(str) => str,
        // TODO: setup a consistent error/normal result format to allow the frontend to read and display potential errors.
        Err(err) => err.to_string()
    }
}

// TODO: make a write_major_file function

fn main() {
    tauri::Builder::default()
        .invoke_handler(tauri::generate_handler![read_major_file])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
