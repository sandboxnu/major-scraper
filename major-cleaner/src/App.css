:root {
  color-scheme: dark;
  overscroll-behavior-x: none;
}

body {
  max-width: 100vw;
  height: 100vh;
  overflow: hidden;
}

#root {
  height: 100%;
}