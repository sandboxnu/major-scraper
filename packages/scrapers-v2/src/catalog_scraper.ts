import axios from "axios";
import * as cheerio from "cheerio";
import {
  CourseListBodyRowType,
  HTMLCatalog,
  HTMLCatalogCourseList,
  HTMLCatalogCourseListBodyRow,
} from "./types";

const loadCatalogHTML = async (url: string): Promise<CheerioStatic> => {
  try {
    const { data } = await axios.get(url);
    const $ = cheerio.load(data);
    return $;
  } catch (error) {
    throw error;
  }
};

export const scrapeMajorDataFromCatalog = async (
  url: string
): Promise<HTMLCatalog> => {
  try {
    const $ = await loadCatalogHTML(url);
    // step 1: Transform scraped table into intermediate representation (IR)
    const transformedScrapedData = transformMajorDataFromCatalog($);

    return transformedScrapedData;
  } catch (error) {
    throw error;
  }
};

const transformMajorDataFromCatalog = ($: CheerioStatic): HTMLCatalog => {
  const majorName: string = $("#site-title").find("h1").text().trim();
  const catalogYear: string = $("#edition").text().split(" ")[0];
  const yearVersion: number = parseInt(catalogYear.split("-")[0]);

  const requirementsContainer = $("#programrequirementstextcontainer");
  const courseLists = transformCourseLists($, requirementsContainer);

  const prgramRequiredHeading = requirementsContainer
    .find("h2")
    .filter(
      (_idx: number, element: CheerioElement) =>
        $(element).text().includes("Program") &&
        $(element).text().includes("Requirement")
    );
  const prgramRequiredHours =
    Number(
      prgramRequiredHeading
        .next()
        .text()
        .split(/[\s\xa0]+/)[0]
    ) || 0;

  return {
    prgramRequiredHours,
    yearVersion,
    majorName,
    courseLists,
  };
};

const transformCourseLists = (
  $: CheerioStatic,
  requirementsContainer: Cheerio
): HTMLCatalogCourseList[] => {
  // use a stack to keep track of the course list title and description
  let descriptions: string[] = [];
  let courseList: HTMLCatalogCourseList[] = [];

  requirementsContainer
    .children()
    .each((_idx: number, element: CheerioElement) => {
      if (element.name.includes("h")) {
        descriptions.push($(element).text());
      } else if (element.name === "table") {
        const tableDesc = descriptions.pop() || "";
        const courseTable = {
          description: tableDesc,
          courseBody: transformCourseListTable($, element),
        };
        courseList.push(courseTable);
      }
      // TODO: scrap all concentrations for business. only applicable to business
      // else if (element.name === 'ul' && $(element).prev().text().includes("concentration")) {
      //   const link = $(element).find("li > a").attr('href');
      // }
    });
  return courseList;
};

const transformCourseListTable = (
  $: CheerioStatic,
  table: CheerioElement
): HTMLCatalogCourseListBodyRow[] => {
  let courseTable: HTMLCatalogCourseListBodyRow[] = [];

  $(table)
    .find("tbody > tr")
    .each((_idx: number, tr: CheerioElement) => {
      // different row type
      const trClass = tr.attribs["class"];
      const firstTdClass = $(tr).children()[0].attribs["class"];

      let courseListBodyRowType: CourseListBodyRowType;
      if (trClass.includes("subheader")) {
        courseListBodyRowType = CourseListBodyRowType.SUBHEADER;
      } else if (trClass.includes("areaheader")) {
        courseListBodyRowType = CourseListBodyRowType.HEADER;
      } else if (trClass.includes("orclass")) {
        courseListBodyRowType = CourseListBodyRowType.OR_COURSE;
      } else if (firstTdClass != null) {
        courseListBodyRowType = CourseListBodyRowType.PLAIN_COURSE;
      } else {
        courseListBodyRowType = CourseListBodyRowType.COMMENT;
      }

      const courseListBodyRow = constructCourseListBodyRow(
        $,
        tr,
        courseListBodyRowType
      );
      courseTable.push(courseListBodyRow);
    });

  return courseTable;
};

const constructCourseListBodyRow = (
  $: CheerioStatic,
  tr: CheerioElement,
  courseListBodyRowType: CourseListBodyRowType
): HTMLCatalogCourseListBodyRow => {
  let hour: number = 0;
  let description: string = "";
  let courseTitle: string = "";
  const courseListRow: {
    [key in keyof HTMLCatalogCourseListBodyRow]: HTMLCatalogCourseListBodyRow[key];
  } = {
    type: courseListBodyRowType,
    hour,
    description,
  };
  $(tr)
    .children()
    .each((_idx: number, td: CheerioElement) => {
      const tdClass = td.attribs["class"];
      if (tdClass != null) {
        if (tdClass.includes("codecol")) {
          if (tdClass.includes("orclass")) {
            hour = parseInt($(tr).prev().children().last().text()) || 0;
          }
          courseTitle = $(td).text();
          courseListRow["courseTitle"] = courseTitle;
        }

        if (tdClass.includes("hourscol")) {
          hour = parseInt($(td).text()) || 0;
        }
      } else {
        description = $(td).text();
      }
    });
  courseListRow["hour"] = hour;
  courseListRow["description"] = description;

  return courseListRow;
};
