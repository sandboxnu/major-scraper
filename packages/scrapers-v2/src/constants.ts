export const COURSE_REGEX = /([A-Z]{2,4})\s([0-9]{4})/g;

export const RANGE_1_REGEX = /([A-Z]{2,4})\s([0-9]{4}) or higher/;
export const RANGE_2_REGEX =
  /([A-Z]{2,4})\s([0-9]{4})(( to )|(-)|(â€“))([A-Z]{2,4})\s([0-9]{4})/;
export const RANGE_3_REGEX =
  /((Select from any)|(Complete \w+)) [A-Z]{2,4} courses? numbered [0-9]{4} or above/;
export const RANGE_3_PARSE_REGEX = /([A-Z]{2,4}) courses? numbered ([0-9]{4})/g;
