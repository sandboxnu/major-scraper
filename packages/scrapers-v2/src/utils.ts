import axios from "axios";
import * as cheerio from "cheerio";
import { CatalogHierarchy } from "./urls/types";

export const loadHTML = async (url: string): Promise<CheerioStatic> => {
  try {
    const { data } = await axios.get(url);
    return cheerio.load(data);
  } catch (error) {
    console.log("request failed:", url);
    throw error;
  }
};

export const appendPath = (base: string, path: string, hash?: string) => {
  const url = new URL(path, base);
  if (hash) {
    url.hash = hash;
  }
  return url;
};

export const ensureLength = <T>(n: number, l: T[]) => {
  const length = l.length;
  if (length !== n) {
    const msg = `expected text row to contain exactly ${n} cells, found ${length}`;
    throw new Error(msg);
  }
  return l;
};

export const ensureLengthAtLeast = <T>(n: number, l: T[]) => {
  const length = l.length;
  if (length < n) {
    const msg = `expected text row to contain at least ${n} cells, found ${length}`;
    throw new Error(msg);
  }
  return l;
};

export const parseText = (td: Cheerio) => {
  // replace &NBSP with space
  return td.text().replaceAll("\xa0", " ").trim();
};

export const flattenCatalogHierarchy = <T extends { url: string }>(
  hierarchy: CatalogHierarchy<T>
): T[] => {
  const list: T[] = [];
  const isLeaf = (node: T | CatalogHierarchy<T>): node is T => {
    // note: this doesn't work if "url" happens to be a key in the hierarchy
    // but that probably won't ever happen... right?
    return "url" in node;
  };
  const recur = (node: T | CatalogHierarchy<T>) => {
    if (isLeaf(node)) {
      // it's a leaf!
      list.push(node);
    } else {
      for (const child of Object.values(node)) {
        recur(child);
      }
    }
  };
  recur(hierarchy);
  return list;
};
