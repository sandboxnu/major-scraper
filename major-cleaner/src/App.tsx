import { useEffect, useState } from "react";
import reactLogo from "./assets/react.svg";
import { invoke } from "@tauri-apps/api/tauri";
import "./App.css";
import { statSync } from "fs";
import { BaseDirectory, readTextFile } from "@tauri-apps/api/fs";
import { Token } from "./components/tokens";

function App() {
  const [greetMsg, setGreetMsg] = useState("");
  const [name, setName] = useState("");
  const [major, setMajor] = useState();
  const [tokens, setTokens] = useState("")

  useEffect(() => {
    readTextFile('bscs-tokens-v1.json', { dir: BaseDirectory.Desktop }).then(val => setTokens(val))
  }, [])

  async function greet() {
    // Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
    setGreetMsg(await invoke("greet", { name }));
  }

  return (
    <div>
      <div style={style.dropdownContainer}>
        <select name="major" id="major" onChange={v => console.log(v.target.value)}>
          <option value="CS">CS</option>
          {/* <option value="saab">Saab</option>
          <option value="mercedes">Mercedes</option>
          <option value="audi">Audi</option> */}
        </select>
      </div>
      <div style={style.container}>
        <div style={style.column}>
          <p>HTML</p>
        </div>
        <div style={style.column}>
          <p>Tokens</p>
          <Token section={JSON.parse(tokens)["tokenized"]["sections"][0]}></Token>

        </div>
        <div style={style.column}>
          <p>Parse</p>
        </div>
      </div>
    </div>
  );
}

const style = {
  column: {
    border: "solid gray",
    flex: 1,
  },
  container: {
    margin: 0,
    paddingTop: "10vh",
    display: "flex",
    flexDirection: "row",
    justifyContent: "flex-start",
    textAlign: "center",
  },
  dropdownContainer: {

  }
  
} satisfies Record<string, React.CSSProperties>;
export default App;
