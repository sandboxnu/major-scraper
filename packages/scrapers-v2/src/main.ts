import { scrapeMajorLinks } from "./urls/scrape_urls";
import { filterByEntryType } from "./filters/filters";
import { CatalogEntryType } from "./filters/types";
import { flattenCatalogHierarchy } from "./utils";
import { fetchAndTokenizeHTML } from "./lexer/lexer";


export const runScrape = async () => {
  try {
    const hierarchy = await scrapeMajorLinks(2021, 2022);
    const majorHierarchy = await filterByEntryType(hierarchy, [CatalogEntryType.Major]);
    const majors = flattenCatalogHierarchy(majorHierarchy);
    const tokenized = await Promise.all(majors.map(({ url }) => fetchAndTokenizeHTML(url)));
    // todo:
    // return await tokenized.map(tok => parseMajor2(tok))
  } catch (e) {
    throw e;
  }
}