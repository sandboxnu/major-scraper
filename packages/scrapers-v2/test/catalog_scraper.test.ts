import { scrapeMajorDataFromCatalog } from "../src/catalog_scraper";

describe("Scrap catalog page", () => {
  test("CS & Game Dev matches snapshot", async () => {
    const result = await scrapeMajorDataFromCatalog(
      "https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-game-development-bs/#programrequirementstext"
    );

    expect(result).toMatchSnapshot();
  });
});
