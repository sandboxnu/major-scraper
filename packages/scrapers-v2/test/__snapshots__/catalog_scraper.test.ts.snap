// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrap catalog page CS & Game Dev matches snapshot 1`] = `
Object {
  "courseLists": Array [
    Object {
      "courseBody": Array [
        Object {
          "description": "Computer Science Overview",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "CS 1200",
          "description": "First Year Seminar",
          "hour": 1,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or ARTF 1000",
          "description": " Art and Design at Northeastern",
          "hour": 1,
          "type": "OR_COURSE",
        },
        Object {
          "courseTitle": "CS 1210",
          "description": "Professional Development for Khoury Co-op",
          "hour": 1,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or EEAM 2000",
          "description": " Professional Development for Co-op",
          "hour": 1,
          "type": "OR_COURSE",
        },
        Object {
          "description": "Computer Science Fundamental Courses",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "CS 1800and CS 1802",
          "description": "Discrete Structuresand Seminar for CS 1800",
          "hour": 5,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 2500and CS 2501",
          "description": "Fundamentals of Computer Science 1and Lab for CS 2500",
          "hour": 5,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 2510and CS 2511",
          "description": "Fundamentals of Computer Science 2and Lab for CS 2510",
          "hour": 5,
          "type": "PLAIN_COURSE",
        },
        Object {
          "description": "Computer Science Required Courses",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "CS 3000",
          "description": "Algorithms and Data",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 3500and CS 3501",
          "description": "Object-Oriented Designand Lab for CS 3500",
          "hour": 5,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 3520",
          "description": "Programming in C++ (Integrative course)",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 3540",
          "description": "Game Programming (Integrative course)",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 3650",
          "description": "Computer Systems",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 3700",
          "description": "Networks and Distributed Systems",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 4300",
          "description": "Computer Graphics (Integrative course)",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "CS 4500",
          "description": "Software Development",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or CS 4530",
          "description": " Fundamentals of Software Engineering",
          "hour": 4,
          "type": "OR_COURSE",
        },
        Object {
          "courseTitle": "CS 4850",
          "description": "Building Game Engines (Integrative course)",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "description": "Computer Science Elective Course",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "CS 4150",
          "description": "Game Artificial Intelligence (Integrative course)",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or IS 4300",
          "description": " Human Computer Interaction",
          "hour": 4,
          "type": "OR_COURSE",
        },
      ],
      "description": "Computer Science Courses",
    },
    Object {
      "courseBody": Array [
        Object {
          "description": "Game Design Required",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "GAME 1110",
          "description": "Games and Society",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "GAME 2500",
          "description": "Foundations of Game Design",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "GAME 2950",
          "description": "Game Studio",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "GAME 3400",
          "description": "Level Design and Game Architecture",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "GAME 3700",
          "description": "Rapid Idea Prototyping for Games",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "GAME 3800",
          "description": "Game Concept Development",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "GAME 4700",
          "description": "Game Design Capstone",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "description": "Game Design Elective",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "GAME 1850",
          "description": "Experimental Game Design",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or GAME 3300",
          "description": " Game Interface Design",
          "hour": 4,
          "type": "OR_COURSE",
        },
        Object {
          "description": "Khoury/Game-Related Electives",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "description": "Complete three of the following:",
          "hour": 12,
          "type": "COMMENT",
        },
        Object {
          "description": "Any course in ARTD, ARTE, ARTF, ARTG, ARTH, and GAME subject areas as long as prerequisites have been met.",
          "hour": 0,
          "type": "COMMENT",
        },
        Object {
          "description": "CS 2500 or higher, except CS 5010",
          "hour": 0,
          "type": "COMMENT",
        },
        Object {
          "description": "CY 2000 or higher, except CY 4930",
          "hour": 0,
          "type": "COMMENT",
        },
        Object {
          "description": "DS 2500 or higher, except DS 4900",
          "hour": 0,
          "type": "COMMENT",
        },
        Object {
          "description": "IS 2000 or higher, except IS 4900",
          "hour": 0,
          "type": "COMMENT",
        },
        Object {
          "courseTitle": "MATH 1342",
          "description": "Calculus 2 for Science and Engineering",
          "hour": 0,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "MATH 2331",
          "description": "Linear Algebra",
          "hour": 0,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "ECON 2350",
          "description": "Statistics",
          "hour": 0,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or PSYC 2320",
          "description": " Statistics in Psychological Research",
          "hour": 0,
          "type": "OR_COURSE",
        },
      ],
      "description": "Game Design Courses",
    },
    Object {
      "courseBody": Array [
        Object {
          "description": "Psychology",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "PSYC 1101",
          "description": "Foundations of Psychology",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "description": "Mathematics",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "description": "Complete two courses from the following:",
          "hour": 8,
          "type": "COMMENT",
        },
        Object {
          "courseTitle": "MATH 1260",
          "description": "Math Fundamentals for Games (Integrative course)",
          "hour": 0,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "MATH 1341",
          "description": "Calculus 1 for Science and Engineering",
          "hour": 0,
          "type": "PLAIN_COURSE",
        },
        Object {
          "description": "MATH 1342 or higher",
          "hour": 0,
          "type": "COMMENT",
        },
      ],
      "description": "Supporting Courses",
    },
    Object {
      "courseBody": Array [
        Object {
          "description": "College Writing",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "courseTitle": "ENGW 1111",
          "description": "First-Year Writing",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "description": "Advanced Writing in the Disciplines",
          "hour": 0,
          "type": "HEADER",
        },
        Object {
          "description": "Students are strongly encouraged to take Advanced Writing in the Technical Professions (ENGW 3302) but may substitute Interdisciplinary Advanced Writing in the Disciplines (ENGW 3315).",
          "hour": 0,
          "type": "COMMENT",
        },
        Object {
          "courseTitle": "ENGW 3302",
          "description": "Advanced Writing in the Technical Professions",
          "hour": 4,
          "type": "PLAIN_COURSE",
        },
        Object {
          "courseTitle": "or ENGW 3315",
          "description": " Interdisciplinary Advanced Writing in the Disciplines",
          "hour": 4,
          "type": "OR_COURSE",
        },
      ],
      "description": "Computer Science Writing Requirement",
    },
    Object {
      "courseBody": Array [
        Object {
          "description": "Complete 12 semester hours of general electives.",
          "hour": 12,
          "type": "COMMENT",
        },
      ],
      "description": "Required General Electives",
    },
  ],
  "majorName": "Computer Science and Game Development, BS",
  "prgramRequiredHours": 133,
  "yearVersion": 2021,
}
`;
